src/box/box.cc

```diff
@@ -1805,6 +1811,228 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
 	return 0;
 }
 
+/** A structure holding trigger data to collect syncs. */
+struct sync_trigger_data {
+	/** Syncs to wait for. */
+	uint64_t vclock_syncs[VCLOCK_MAX];
+	/**
+	 * A bitmap holding replica ids whose vclocks were already collected.
+	 */
+	vclock_map_t collected_vclock_map;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+	/** Collected vclock. */
+	struct vclock *vclock;
+	/** The request deadline. */
+	double deadline;
+	/** How many vclocks are needed. */
+	int count;
+	/** Whether the request is timed out. */
+	bool is_timed_out;
+};
+
+/**
+ * A trigger executed on each ack to collect up to date remote node vclocks.
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_vclock_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->vclock_syncs[id];
+	int accounted_count = bit_count_u32(data->collected_vclock_map) + 1;
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->vclock_sync >= sync && accounted_count < data->count) {
```

Comment:
> Why isn't it `<= data->count` here

Because if it's equal to `data->count`, the waiter was already woken up on the previous iteration.

> and why do you increment accounted_count below even though it is already +1 on the line 1847

It's +1 on line 1847, because I count the local node separately.
+1 below is for the freshly accounted vclock.

Anyway, it seems I better account the local vclock in the map as well. Otherwise the code's rather complicated.
Will paste the diff here once the rework is done.

Reply:
Here:
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 01318c9cd..8c843ec5f 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1843,8 +1843,14 @@ check_vclock_sync_on_ack(struct trigger *trigger, void *event)
 	struct sync_trigger_data *data =
 		(struct sync_trigger_data *)trigger->data;
 	uint32_t id = ack->source;
+	/*
+	 * Anonymous replica acks are not counted for synchronous transactions,
+	 * so linearizable read shouldn't count them as well.
+	 */
+	if (id == 0)
+		return 0;
 	uint64_t sync = data->vclock_syncs[id];
-	int accounted_count = bit_count_u32(data->collected_vclock_map) + 1;
+	int accounted_count = bit_count_u32(data->collected_vclock_map);
 	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
 	    ack->vclock_sync >= sync && accounted_count < data->count) {
 		vclock_max_ignore0(data->vclock, ack->vclock);
@@ -1906,6 +1912,7 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 		.count = vclock_count,
 		.is_timed_out = false,
 	};
+	bit_set(&data.collected_vclock_map, instance_id);
 	struct trigger on_relay_thread_start;
 	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
 		       NULL);
@@ -1931,7 +1938,7 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 		}
 	}
 
-	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
+	while (bit_count_u32(data.collected_vclock_map) < vclock_count &&
 	       !data.is_timed_out && !fiber_is_cancelled()) {
 		if (fiber_yield_deadline(deadline))
 			break;
@@ -1943,7 +1950,7 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 		diag_set(FiberIsCancelled);
 		return -1;
 	}
-	if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
+	if (bit_count_u32(data.collected_vclock_map) < vclock_count) {
 		diag_set(TimedOut);
 		return -1;
 	}

```
</details>